### ISSUE TITLE
HTML entities should not be auto-converted to characters

### ISSUE THREAD

**Issue by salty-horse** [wordpress-mobile/AztecEditor-iOS#913]
HTML entities entered in HTML mode are automatically converted by Visual mode to the characters they represent.

This can cause problems: When entering `&#xfeff` in HTML mode (ZERO WIDTH NO-BREAK SPACE), switching to the visual mode, then back to the HTML mode, `&#xfeff` turns into the literal U+FEFF Unicode character, making it hard to edit (as it's invisible).

HTML entities inputted in HTML mode should be kept as-is.

**Comment by diegoreymendez** [comment]
I agree that for certain characters (for example invisible ones) it may make sense to show their entity representation in HTML mode.  Thus `&#xfeff` should always be shown as `&#xfeff` whenever you switch to HTML mode.

That said, this issue is broader than just that, judging from the title and description.

Just to clarify, Aztec does not maintain a perfect 1:1 representation of the HTML input when switching modes to visual and back, and that's an architectural decision to avoid excessive (and really not necessary) complexity.

In other words: the representation of any specific character when switching to HTML mode should be based on a predefined mapping table rather than on how that character was inserted while in HTML mode.

**Comment by salty-horse** [comment]
The impetus for this bug report was this zero-width character, but I'm wary of any decision the app makes for the user, even if it's a small mapping table, and so I framed it as a larger issue.

If the user has intentionally input a character in HTML mode as an entity, rather than a printable character, they might have a reason for it, and want it kept that way. For example, if they want to differentiate em-dash (`&mdash;`) and en-dash (`&ndash;`), which look very similar.
I'm all for making the user's workflow easier in visual mode, but HTML mode is "I know what I'm doing".

If it's technically complicated and clutters the codebase, sure, do the right thing for the editor.

**Comment by diegoreymendez** [comment]
In simpler terms, the architectural direction is:

```The editor normalizes the output HTML in its own way, as long as it doesn't change the way the content is rendered.```

One such example is node ordering.  These inputs:

```html
<em><strong>Hello</strong> world</em>
```

```html
<strong><em>Hello</em></strong><em> world</em>
```

Are both converted to:

```html
<p><em><strong>Hello</strong> world</em></p>
```

One exception to this rule is CSS attributes (which still need tons of work in Aztec).  Because while we're not really rendering much from CSS, these attributes contain information which may affect how the content is rendered online.

> If the user has intentionally input a character in HTML mode as an entity, rather than a printable character, they might have a reason for it, and want it kept that way.

While I realize the intent is noble behind this, I'd be wary to consider this as necessary unless we find out it really is through some specific use case.

For the time being I'd narrow this issue down to any specific characters that should definitely be mapped to its entity (we can even make this mapping table customizable by the App, if we want to).

### Additional Background:

Before picking this architectural direction we went through a much more complicated engine that would ensure a perfect 1:1 input vs output, but had to change route since it was becoming extremely complex.

@jleandroperez and I ended up realizing that we were putting a lot of work into a theoretical need to maintain this equivalence, and coming up with a really complex product that would be hard to extend maintain.  In the end, when typing characters in visual mode, you don't get to decide if they're converted to entities or not, so having an App setting for this (even one that can be configured by the user) seems more appropriate than trying to maintain the input.

**Comment by salty-horse** [comment]
Thanks for the explanation! I agree with this approach, although I can't imagine how destructive it is (if at all) in its current state to existing pages that make use of CSS. Hopefully it doesn't delete anything for existing users. But this is getting off topic.

If I understand your meaning about an app setting, I don't think one that controls how the visual mode's output HTML looks (in terms of  HTML entities) is useful to anyone. It's also an unknown/confusing concept to the average user. Select conversion of some entities (either using a hard-coded table, or a function based on Unicode ranges and/or properties) might be the best approach.

**Comment by diegoreymendez** [comment]
> It's also an unknown/confusing concept to the average user. Select conversion of some entities (either using a hard-coded table, or a function based on Unicode ranges and/or properties) might be the best approach.

Exactly what's in my thoughts.

I think either hard-coded or one that can be configured by the client App (not necessarily the user though).

**Comment by diegoreymendez** [comment]
Following the discussion above, I'm closing this issue in favor of [a more specific / atomic one](https://github.com/wordpress-mobile/AztecEditor-iOS/issues/914).

We can create specific issued for any other mappings
